src/interpreter/plugin/FinancialPlugin.ts

```diff
@@ -559,6 +599,118 @@ export class FinancialPlugin extends FunctionPlugin {
+    return this.runFunction(ast.args, formulaAddress, this.metadata('FVSCHEDULE'),
+      (value: number, ratios: SimpleRangeValue) => {
+        const vals = ratios.valuesFromTopLeftCorner()
+        for(const val of vals) {
+          if(val instanceof CellError) {
+            return val
+          }
+        }
+        for(const val of vals) {
+          if(typeof val === 'number') {
+            value *= 1+val
+          } else if(val !== EmptyValue) {
+            return new CellError(ErrorType.VALUE, ErrorMessage.NumberExpected)
+          }
+        }
+        return value
+      })
+  }
+
+  public npv(ast: ProcedureAst, formulaAddress: SimpleCellAddress): InternalScalarValue {
+    return this.runFunction(ast.args, formulaAddress, this.metadata('NPV'),
+      (rate: number, ...args: InterpreterValue[]) => {
+        const coerced = this.interpreter.arithmeticHelper.coerceNumbersExpandRanges(args)
+        if(coerced instanceof CellError) {
+          return coerced
+        }
+        return npvCore(rate, coerced)
+      }
+    )
+  }
+
+  public mirr(ast: ProcedureAst, formulaAddress: SimpleCellAddress): InternalScalarValue {
+    return this.runFunction(ast.args, formulaAddress, this.metadata('MIRR'),
+      (range: SimpleRangeValue, frate: number, rrate: number) => {
+        const vals = this.interpreter.arithmeticHelper.manyToExactNumbers(range.valuesFromTopLeftCorner())
+        if(vals instanceof CellError) {
+          return vals
+        }
+        let posFlag = false
+        let negFlag = false
+        const posValues: number[] = []
+        const negValues: number[] = []
+        for(const val of vals) {
+          if(val>0) {
+            posFlag = true
+            posValues.push(val)
+            negValues.push(0)
+          } else if(val<0) {
+            negFlag = true
+            negValues.push(val)
+            posValues.push(0)
+          } else {
+            negValues.push(0)
+            posValues.push(0)
+          }
+        }
+        if(!posFlag || !negFlag) {
+          return new CellError(ErrorType.DIV_BY_ZERO)
+        }
+        const n = vals.length
+        return Math.pow(
+          (-npvCore(rrate, posValues)*Math.pow(1+rrate, n)/npvCore(frate, negValues)/(1+frate)),
+          1/(n-1)
+        )-1
+      }
+    )
+  }
+
+  public pduration(ast: ProcedureAst, formulaAddress: SimpleCellAddress): InternalScalarValue {
+    return this.runFunction(ast.args, formulaAddress, this.metadata('PDURATION'),
+      (rate: number, pv: number, fv: number) => (Math.log(fv) - Math.log(pv))/Math.log(1+rate)
+    )
+  }
+
+  public xnpv(ast: ProcedureAst, formulaAddress: SimpleCellAddress): InternalScalarValue {
+    return this.runFunction(ast.args, formulaAddress, this.metadata('XNPV'),
+      (rate: number, values: SimpleRangeValue, dates: SimpleRangeValue) => {
+        const valArr = values.valuesFromTopLeftCorner()
+        for(const val of valArr) {
```

Comment:
don't we have to check that there is at least one positive and one negative value?

Reply:
in XL docs: yes. in XL implementation: no